LION/optimizers/ARsolver.py

```diff
@@ -0,0 +1,283 @@
+                    g["lr"] = lr
+
+                optimizer.step()
+
+            validation_loss = np.append(
+                validation_loss, self.validation_fn(recon, target).item()
+            )
+
+        if self.verbose:
+            print(
+                f"Validation loss mean: {validation_loss.mean()} - Validation loss std: {validation_loss.std()}"
+            )
+
+        # return to train if it was in train
+        if status:
+            self.model.train()
+
+        return np.mean(validation_loss)
+
+    def test(self):
+        if self.check_testing_ready() != 0:
+            raise LIONSolverException(
+                "Solver not ready for testing. Please call set_testing."
+            )
+
+        # these always pass if the above does, this is just to placate static type checker
+        assert self.test_loader is not None
+        assert self.testing_fn is not None
+
+        status = self.model.training
+        self.model.eval()
+
+        if self.verbose:
+            print(f"Testing model after {self.current_epoch} epochs of training")
+
+        self.estimate_alpha(self.test_loader)
+
+        test_loss = np.array([])
+        for sino, target in tqdm(self.test_loader):
+            bad_recon = fdk(sino, self.op)
+            # is it worth normalizing in testing?
+            # if self.do_normalize:
+            #     bad_recon = self.normalize(bad_recon)
+            #     target = self.normalize(target)
+
+            recon = torch.nn.Parameter(bad_recon.clone()).requires_grad_(True)
+
+            optimizer = self.recon_optimizer_type(
+                [recon],
+                *self.solver_params.optimizer_params.unpack(),
+                lr=self.solver_params.step_size,
+            )
+            lr = self.solver_params.step_size
+
+            for _ in tqdm(range(self.solver_params.no_steps)):
+                optimizer.zero_grad()
+
+                energy = self.var_energy(recon, sino)
+                energy.backward(retain_graph=True)
+
+                assert recon.grad is not None
+                while (
+                    self.var_energy(recon - recon.grad * lr, sino)
+                    > energy - 0.5 * lr * (recon.grad.norm(dim=(2, 3)) ** 2).mean()
+                ):
+                    lr = self.solver_params.beta_rate * lr
+                for g in optimizer.param_groups:
+                    g["lr"] = lr
+
+                optimizer.step()
+
+                with torch.no_grad():
+                    test_loss = np.append(
+                        test_loss, self.testing_fn(recon, target).item()
+                    )
+
+        if self.verbose:
+            print(
+                f"Testing loss mean: {test_loss.mean()} - Testing loss std: {test_loss.std()}"
+            )
+
+        # return to train if it was in train
+        if status:
+            self.model.train()
+
+        return test_loss
+
+    def wgan_loss(self, sino_batch, target_batch):
```

Comment:
@Zakobian is "WGAN-loss" a name _too generic_? It is _a_ WGAN loss, of course, but would someone mistake this for something else? or I am overworrying about silly things?

Reply:
Id say this is overworrying, cause it is the WGAN loss

Reply:
leave as is then @AnderBiguri